### PR TITLE
Trace forecasts

### PR DESCRIPTION
**Pull request by kilianp14**
- A bit of restructuring of core.simulator so that some functionality is abstracted from CarbonApi
- Integration of a forecast_at function to get forecasts: A frequency and a resample_method can be specified and data will be interpolated accordingly (detailed information about the functionality can be found in the docstring). Tests for this function are implemented too
- Auto-formatting of the test files
- Implementation of apply_error function for the datasets (some changes need to be made though #142)

### CODE REVIEW

File: vessim/core/simulator.py [hunk 1]
```diff
@@ -8,75 +8,141 @@
...
+            ------------------------------------------------------------------------
+            Timestamp                                   Zone A  Zone B  Zone C
+                                  2020-01-01 01:00:00   115     850     710
+            2020-01-01 00:00:00   2020-01-01 02:00:00   120     850     715
+                                  2020-01-01 03:00:00   110     870 	720
+
+                                  2020-01-01 02:00:00   110     830     715
+            2020-01-01 01:00:00   2020-01-01 03:00:00   105     860     725
+                                  2020-01-01 04:00:00   120     870 	740
+            ------------------------------------------------------------------------
```
Comment (marvin-steinke) [inline vessim/core/simulator.py:34]: why are the forecasts formatted differently than the actual data?

File: vessim/core/simulator.py [hunk 2]
```diff
@@ -8,75 +8,141 @@
...
+                raise ValueError("Zone needs to be specified.")
+        try:
+            if self.forecast is None:
+                zone_forecast = self.actual[zone][self.actual.index >= dt]
+            else:
+                zone_forecast = self.forecast[zone]
+        except KeyError:
+            raise ValueError(f"Cannot retrieve forecast at zone '{zone}'.")
+        try:
+            return zone_forecast.loc[self.data.index.asof(dt)]
```
Comment (marvin-steinke) [inline vessim/core/simulator.py:88]:
```suggestion
            return zone_forecast.loc[self.actual.index.asof(dt)]
```

File: vessim/core/simulator.py [hunk 3]
```diff
@@ -8,75 +8,141 @@
...
         """Returns the next time of when the trace will change.
 
         This method is being called in the time-based simulation model for Mosaik.
         """
-        current_index = self.data.index.asof(dt)
-        next_iloc = self.data.index.get_loc(current_index) + 1
-        return self.data.index[next_iloc]
+        current_index = self.actual.index.asof(dt)
+        next_iloc = self.actual.index.get_loc(current_index) + 1
+        return self.actual.index[next_iloc]
```
Comment (marvin-steinke) [inline vessim/core/simulator.py:99]: Are out of bounds exceptions an issue here?

File: vessim/core/simulator.py [hunk 4]
```diff
@@ -8,75 +8,141 @@
...
+            if self.forecast is not None:
+                self.forecast = self.forecast * 0.45359237
         elif unit != "g_per_kWh":
             raise ValueError(f"Carbon intensity unit '{unit}' is not supported.")
 
-    def zones(self) -> List:
-        """Returns a list of all available zones."""
-        return list(self.data.columns)
-
     def carbon_intensity_at(self, dt: Time, zone: Optional[str] = None) -> float:
```
Comment (marvin-steinke) [inline vessim/core/simulator.py:128]: Do we even need this and the `power_at` function anymore?

**Review by marvin-steinke — CHANGES_REQUESTED** [review]
Looking good!

**Review by birnbaum — APPROVED** [review]

**Pull request merged**